[DoNotMerge] Introducing subAssessments type (Microsoft.Security provider)

orparnes opened the pull request:
### Latest improvements:
See also [#5649 - Introducing assessments and assessmentsMetadata types](https://github.com/Azure/azure-rest-api-specs/pull/5649)

<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/azure-rest-api-specs#basics) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [x] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [x] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [ ] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

orparnes committed:
Add sub-assessment resource to Microsoft.Security provider

AutorestCI commented:
# Automation for azure-sdk-for-python

azuresdkci commented:
Can one of the admins verify this patch?

AutorestCI commented:
# Automation for azure-sdk-for-go

orparnes commented on specification/security/resource-manager/Microsoft.Security/preview/2019-01-01-preview/subassessments.json:
```diff
@@ -0,0 +1,545 @@
+				"type": {
+					"readOnly": true,
+					"type": "string",
+					"description": ""
```
>"description": "" [](start = 5, length = 17)

all empty descriptions - TBD

AutorestCI commented:
# Automation for azure-sdk-for-ruby

AutorestCI commented:
# Automation for azure-sdk-for-java

orparnes commented on specification/security/resource-manager/Microsoft.Security/preview/2019-01-01-preview/examples/SubAssessments/ListSubAssessments_example.json:
```diff
@@ -0,0 +1,59 @@
+{
+	"parameters": {
+		"api-version": "2019-01-01-preview",
+		"scope": "subscriptions/20ff7fc3-e762-44dd-bd96-b71116dcdc23"
```
>scope [](start = 3, length = 5)

subAssessment scope is always the parent assessment

should be represented here (?)

nschonni commented on specification/security/resource-manager/Microsoft.Security/preview/2019-01-01-preview/subassessments.json:
```diff
@@ -0,0 +1,545 @@
+					"description": "User friendly display name of the sub-assessment"
+				},
+				"status": {
+					"$ref": "#/definitions/SubAssessementStatus"
```
Assessement -> Assessment
```suggestion
					"$ref": "#/definitions/SubAssesementStatus"
```

nschonni commented on specification/security/resource-manager/Microsoft.Security/preview/2019-01-01-preview/subassessments.json:
```diff
@@ -0,0 +1,545 @@
+				}
+			}
+		},
+		"SubAssessementStatus": {
```
```suggestion
		"SubAssessmentStatus": {
```

nschonni commented on specification/security/resource-manager/Microsoft.Security/preview/2019-01-01-preview/subassessments.json:
```diff
@@ -0,0 +1,545 @@
+							}
+						]
+					},
+					"description": "The sub-assessement severity level"
```
```suggestion
					"description": "The sub-assessment severity level"
```

nschonni commented on specification/security/resource-manager/Microsoft.Security/preview/2019-01-01-preview/subassessments.json:
```diff
@@ -0,0 +1,545 @@
+					"type": "string",
+					"description": "Description of the impact of resolving this sub-assessment"
+				},
+				"catagory": {
```
```suggestion
				"category": {
```

nschonni commented on specification/security/resource-manager/Microsoft.Security/preview/2019-01-01-preview/subassessments.json:
```diff
@@ -0,0 +1,545 @@
+				"catagory": {
+					"readOnly": true,
+					"type": "string",
+					"description": "Catagory of the sub-assessment"
```
```suggestion
					"description": "Category of the sub-assessment"
```

nschonni commented on specification/security/resource-manager/Microsoft.Security/preview/2019-01-01-preview/subassessments.json:
```diff
@@ -0,0 +1,545 @@
+					"readOnly": true,
+					"type": "object",
+					"description": "Array of CVSS. Key is CVSS version",
+					"additionalproperties":
```
```suggestion
					"additionalProperties":
```

nschonni commented on specification/security/resource-manager/Microsoft.Security/preview/2019-01-01-preview/subassessments.json:
```diff
@@ -0,0 +1,545 @@
+					"readOnly": true,
+					"type": "object",
+					"description": "Array of CVSS. Key is CVSS version",
+					"additionalproperties":
```
```suggestion
					"additionalProperties":
```

nschonni commented on specification/security/resource-manager/Microsoft.Security/preview/2019-01-01-preview/subassessments.json:
```diff
@@ -0,0 +1,545 @@
+				"repositoryName": {
+					"readOnly": true,
+					"type": "string",
+					"description": "Name of the repository which the vurnerable image belongs to"
```
```suggestion
					"description": "Name of the repository which the vulnerable image belongs to."
```

nschonni commented on specification/security/resource-manager/Microsoft.Security/preview/2019-01-01-preview/subassessments.json:
```diff
@@ -0,0 +1,545 @@
+				"imageName": {
+					"readOnly": true,
+					"type": "string",
+					"description": "Name of the vurnerable image"
```
```suggestion
					"description": "Name of the vulnerable image."
```

nschonni commented on specification/security/resource-manager/Microsoft.Security/preview/2019-01-01-preview/subassessments.json:
```diff
@@ -0,0 +1,545 @@
+				"imageDigest": {
+					"readOnly": true,
+					"type": "string",
+					"description": "Digest of the vurnerable image"
```
```suggestion
					"description": "Digest of the vulnerable image."
```

nschonni reviewed (commented):
CVSS can be added to https://github.com/Azure/azure-rest-api-specs/blob/master/custom-words.txt to ignore it from the spellcheck

orparnes committed:
Fix typos, Add custom words, Descriptions, Renames, Examples

nschonni commented on specification/security/resource-manager/Microsoft.Security/preview/2019-01-01-preview/subAssessments.json:
```diff
@@ -0,0 +1,556 @@
+				"timeGenerated": {
+					"readOnly": true,
+					"type": "string",
+					"description": "The date and time the sub-assessemnt was generated. The format is ISO-8601 (UTC)"
```
PS: Think there may be a "format" way to express the date
```suggestion
					"description": "The date and time the sub-assessment was generated. The format is ISO-8601 (UTC)."
```

nschonni commented on specification/security/resource-manager/Microsoft.Security/preview/2019-01-01-preview/subAssessments.json:
```diff
@@ -0,0 +1,556 @@
+				"imageDigest": {
+					"readOnly": true,
+					"type": "string",
+					"description": "Digest of the vurnerable image"
```
```suggestion
					"description": "Digest of the vulnerable image."
```

nschonni reviewed (commented):
PS: Think there may be a "format" way to express the date. `"format": "date-time"`

nschonni commented on specification/security/resource-manager/Microsoft.Security/preview/2019-01-01-preview/subAssessments.json:
```diff
@@ -0,0 +1,547 @@
+					"readOnly": true,
+					"type": "string",
+					"format": "date-time",
+					"description": "The date and time the sub-assessement was generated"
```
```suggestion
					"description": "The date and time the sub-assessment was generated."
```

nschonni commented on specification/security/resource-manager/Microsoft.Security/preview/2019-01-01-preview/subAssessments.json:
```diff
@@ -0,0 +1,547 @@
+				"assessedResourceType": {
+					"readOnly": true,
+					"type": "string",
+					"description": "Sub-assessement resource type",
```
```suggestion
					"description": "Sub-assessment resource type.",
```

nschonni commented on custom-words.txt:
```diff
@@ -310,6 +310,8 @@ customvision
 customvisionprediction
 customvisiontraining
 CUTOVER
+CVE
```
```suggestion
CVE
CVEs
```

nschonni commented on specification/security/resource-manager/Microsoft.Security/preview/2019-01-01-preview/examples/SubAssessments/GetSubAssessment_example.json:
```diff
@@ -0,0 +1,60 @@
+						"Cause": ""
+					},
+					"resourceDetails": {
+						"Source": "Azure",
```
```suggestion
						"source": "Azure",
```

nschonni commented on specification/security/resource-manager/Microsoft.Security/preview/2019-01-01-preview/examples/SubAssessments/ListSubAssessments_example.json:
```diff
@@ -0,0 +1,60 @@
+							"Cause": ""
+						},
+						"resourceDetails": {
+							"Source": "Azure",
```
```suggestion
							"source": "Azure",
```

nschonni commented on specification/security/resource-manager/Microsoft.Security/preview/2019-01-01-preview/examples/SubAssessments/GetSubAssessment_example.json:
```diff
@@ -0,0 +1,60 @@
+						"type": "Vulnerability",
+						"cvss": {
+							"2.0": {
+								"base": "10"
```
```suggestion
								"base": 10
```

nschonni commented on specification/security/resource-manager/Microsoft.Security/preview/2019-01-01-preview/examples/SubAssessments/GetSubAssessment_example.json:
```diff
@@ -0,0 +1,60 @@
+								"base": "10"
+							},
+							"3.0": {
+								"base": "10"
```
```suggestion
								"base": 10
```

nschonni commented on specification/security/resource-manager/Microsoft.Security/preview/2019-01-01-preview/examples/SubAssessments/GetSubAssessment_example.json:
```diff
@@ -0,0 +1,60 @@
+					"displayName": "'Back Orifice' Backdoor",
+					"id": "1001",
+					"status": {
+						"Severity": "High",
```
```suggestion
						"severity": "High",
```

nschonni commented on specification/security/resource-manager/Microsoft.Security/preview/2019-01-01-preview/examples/SubAssessments/GetSubAssessment_example.json:
```diff
@@ -0,0 +1,60 @@
+					"id": "1001",
+					"status": {
+						"Severity": "High",
+						"Cause": ""
```
```suggestion
						"cause": ""
```

nschonni commented on specification/security/resource-manager/Microsoft.Security/preview/2019-01-01-preview/examples/SubAssessments/ListSubAssessments_example.json:
```diff
@@ -0,0 +1,60 @@
+						"displayName": "'Back Orifice' Backdoor",
+						"id": "1001",
+						"status": {
+							"Severity": "High",
```
```suggestion
							"severity": "High",
```

nschonni commented on specification/security/resource-manager/Microsoft.Security/preview/2019-01-01-preview/examples/SubAssessments/ListSubAssessments_example.json:
```diff
@@ -0,0 +1,60 @@
+						"id": "1001",
+						"status": {
+							"Severity": "High",
+							"Cause": ""
```
```suggestion
							"cause": ""
```

nschonni commented on specification/security/resource-manager/Microsoft.Security/preview/2019-01-01-preview/examples/SubAssessments/ListSubAssessments_example.json:
```diff
@@ -0,0 +1,60 @@
+							"type": "Vulnerability",
+							"cvss": {
+								"2.0": {
+									"base": "10"
```
```suggestion
									"base": 10
```

nschonni commented on specification/security/resource-manager/Microsoft.Security/preview/2019-01-01-preview/examples/SubAssessments/ListSubAssessments_example.json:
```diff
@@ -0,0 +1,60 @@
+									"base": "10"
+								},
+								"3.0": {
+									"base": "10"
```
```suggestion
									"base": 10
```

nschonni reviewed (commented):
Mostly camelCase stuff

orparnes commented on specification/security/resource-manager/Microsoft.Security/preview/2019-01-01-preview/subAssessments.json:
```diff
@@ -0,0 +1,547 @@
+					"readOnly": true,
+					"type": "string",
+					"description": "The platform where the assessed resource resides",
+					"enum": ["Azure", "OnPremise"],
```
>Azure [](start = 15, length = 5)

?

orparnes committed:
Add: ResourceDetails, custom words. Fix: examples, typos.

orparnes committed:
Rename "extendedData" to "additionalData" to be aligned with "assessment" type, fix examples, add field, edit status object

ravbhatnagar commented on specification/security/resource-manager/Microsoft.Security/preview/2019-01-01-preview/subAssessments.json:
```diff
@@ -0,0 +1,604 @@
+				}
+			}
+		},
+		"/{resourceId}/providers/Microsoft.Security/assessments/{assessmentName}/subAssessments/{subAssessmentName}": {
```
please use same parameter as the API above. so you can call it {scope}

orparnes replied:
@chlahav We talked about it a bit F2F.
Can you clarify?

sanjaiganesh replied:
replace {resourceId} with  {scope} as Gaurav suggested
{
      "$ref": "../../../common/v1/types.json#/parameters/Scope"
     },

---
In reply to: [299172776](https://github.com/Azure/azure-rest-api-specs/pull/6422#discussion_r299172776) [](ancestors = 299172776)

ravbhatnagar commented on specification/security/resource-manager/Microsoft.Security/preview/2019-01-01-preview/subAssessments.json:
```diff
@@ -0,0 +1,604 @@
+		}
+	},
+	"paths": {
+		"/{scope}/providers/Microsoft.Security/assessments/{assessmentName}/subAssessments": {
```
how is an assessment different from a sub-assessment? It may be better to have a quick 30 min sync to go over this and close. Is this already live on the service side?

orparnes replied:
Had conversion with @sanjaiganesh :)

sanjaiganesh replied:
@  you may add a summary for all.. example + reason

---
In reply to: [300073944](https://github.com/Azure/azure-rest-api-specs/pull/6422#discussion_r300073944) [](ancestors = 300073944)

ravbhatnagar reviewed (commented):
@orparnes - please set up a 30 min call to close on this. Its not clear whether subassessment is the right model as a nested type. so need to understand the usecases a little better.

orparnes committed:
Renames

sanjaiganesh commented on specification/security/resource-manager/Microsoft.Security/preview/2019-01-01-preview/subAssessments.json:
```diff
@@ -0,0 +1,604 @@
+			"description": "Vendor reference",
+			"properties": {
+				"title": {
+					"readOnly": true,
```
>					 [](start = 0, length = 5)

nit - convert tabs to spaces

orparnes committed:
Resolve some issues, set spaces instead of tabs,
Add fields to ContainerRegistryVulnerabilityProperties and ServerVulnerabilityProperties

sanjaiganesh commented on specification/security/resource-manager/Microsoft.Security/preview/2019-01-01-preview/subAssessments.json:
```diff
@@ -0,0 +1,604 @@
+					"$ref": "#/definitions/SubAssessmentStatus"
+					
+				},
+				"remediation": {
```
>"remediation": { [](start = 4, length = 16)

it would be great if we could provide (aka.ms) link to docs that has details either as part of one of the properties or as separate uri property

orparnes replied:
Not sure I understood this point. "remediation" is data we're getting from the sub-assessment provider and it is a dynamic string instructions on how to remediate the security issue.
Although it could be a url, it won't necessarily be one.

sanjaiganesh commented on specification/security/resource-manager/Microsoft.Security/preview/2019-01-01-preview/subAssessments.json:
```diff
@@ -0,0 +1,604 @@
+					"format": "date-time",
+					"description": "The date and time the sub-assessment was generated"
+				},
+				"resourceDetails": {
```
>resourceDetails [](start = 5, length = 15)

nit nit - assessedResource ?

sanjaiganesh replied:
@orparnes added few comments

orparnes replied:
> > resourceDetails [](start = 5, length = 15)
> 
> nit nit - assessedResource ?

We want to be aligned with "assessments" API, so prefer to leave it as-is

orparnes committed:
Fix example

orparnes committed:
x-ms-enum fix

orparnes committed:
Remove some fields

shahabhijeet commented:
@orparnes please resolve merging conflicts

orparnes committed:
Merge branch 'master' into master

yungezz commented:
hi @orparnes what's your plan on this "DoNotMerge" PR? it's been there 2 months. Is it ok to merge it now? or how about close it for now and reopen if ready for merge?

nschonni commented:
Looks like this got replaced by https://github.com/Azure/azure-rest-api-specs/pull/7101
